### ISSUE TITLE
Enhance "Checking package integrity"

### ISSUE THREAD

**Issue by DmitrySokolov** [conan-io/conan#1207]
When a package contains ~10k files the "Checking package integrity" phase lasts about 10 minutes. 
Is it possible to speed up this phase by checking a file modification time or something like to track untouched files?

**Comment by lasote** [comment]
We have to introduce some benchmarks to know if there is a way to improve it. About the untouched files by time, I think it's not enough secure. Let's see where is the bottle neck and see if we can do anything.

**Comment by memsharded** [comment]
I started to run some benchmarks, md5 computation is the thing taking more time (as expected). FYI, it seems that python (or the underlying md5 computation library) is somewhat caching requests (even between calls, so it might be somewhere on disk) so subsequents calls to the md5 computation are almost instanteneous. Or maybe, the file contents are being cached in memory (by the OS), so the expensive operation is reading them from disk, not computing the md5.

**Comment by memsharded** [comment]
Windows was indeed caching files in memory. Resetting mem with https://technet.microsoft.com/en-us/sysinternals/ff700229.aspx (or just re-booting), makes experiments reproducible. I have profiled:
- The md5 computation is very fast, just a small fraction (<10%) of time.
- The expensive operation is reading the files. For about 6K source files it can take 30secs.
- Modifying ``_generic_algorithm_sum``, increasing read size, or even doing one step ``read()`` doesn't improve things.
So, improvements over the current flow seems difficult.

Reading and comparing files timestamps seems fragile to me. I would say that the whole check could be reconsidered: why this check? protection against mistakes? against reproducibility (users tampering with package contents out of the conan flow)?

**Comment by DmitrySokolov** [comment]
On macOS:
```
Uploading package 1/2: 99d469546e717a7f932160a7ef77c82444f38739
Checking package integrity...                                         
DEBUG :remote_manager.py[69]: ====> Time remote_manager build_files_set : 0.567939 [2017-04-18 00:10:36,645]
DEBUG :remote_manager.py[69]: ====> Time remote_manager build_files_set : 0.567939 [2017-04-18 00:10:36,645]
Package integrity OK!                                                 
DEBUG :remote_manager.py[93]: ====> Time remote_manager check package integrity : 352.459032 [2017-04-18 00:16:28,536]
```

**Comment by memsharded** [comment]
Thanks Dmitry for the data. I think it is aligned with my observations, the initial times are just for gathering the list (not reading) of files to be packaged. The package integrity time (352secs) are for reading the file contents from disk (>90%) of the time and computing their md5.

**Comment by tru** [comment]
I would appriciate a option to opt-out of this check if there is no way to get it to be faster.

**Comment by memsharded** [comment]
Maybe even consider making it opt-in, lets review the value of this feature.

**Comment by memsharded** [comment]
It seems we can make it opt-in, being the default not doing such checks, they should be unnecessary for 99,99% of cases. Proposal in #1454

**Comment by memsharded** [comment]
Released in 0.25, as opt-in.
